### PR TITLE
Add DREAM training

### PR DESCRIPTION
**Pull request by AmericanPresidentJimmyCarter**
# What does this PR do?

A new function compute_dream_and_update_latents has been added to the training utilities that allows you to do DREAM rectified training in line with the paper https://arxiv.org/abs/2312.00210. The method can be used with an extra argument in the train_text_to_image.py script.

Ported from @RossM's webui MR with help from @RossM.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests? (Shouldn't be needed as I think most training_utils.py functions do not have them)

## Who can review?
@sayakpaul and @patrickvonplaten

### CODE REVIEW

**Comment by sayakpaul** [comment]
Can we see some results first?

File: src/diffusers/training_utils.py
Comment (sayakpaul) [inline src/diffusers/training_utils.py:57]: Better would be to return the expected variables rather than doing it in-place. It's more readable that way in a long training script.

Comment (AmericanPresidentJimmyCarter) [reply src/diffusers/training_utils.py:57]: Sure.

**Comment by AmericanPresidentJimmyCarter** [comment]
> Can we see some results first?

Yeah don't merge. It works for eps pred but I am not sure @rossm has figured out v-pred (as written it does not work), so I may just remove that and throw an error for v-pred.

**Comment by AmericanPresidentJimmyCarter** [comment]
292780312-beff2145-1499-406e-9cce-da3635cbd997

normal eps pred finetuning on left, eps pred with DREAM finetuning on the right

**Comment by sayakpaul** [comment]
Results look nice. Some questions:

* Assuming both the above runs had the exact same params?
* Did you notice any hyperparameter sensitivity when using DREAM? Like did you have change anything particular non-DREAM CLI args which you didn't change for the experiment without DREAM?
* How about DreamBooth + LoRA?

**Comment by AmericanPresidentJimmyCarter** [comment]
> Results look nice. Some questions:
> 
>     * Assuming both the above runs had the exact same params?
yes

>     * Did you notice any hyperparameter sensitivity when using DREAM? Like did you have change anything particular non-DREAM CLI args which you didn't change for the experiment without DREAM?
no

>     * How about DreamBooth + LoRA?
Have not tried

>     * What are the memory implications?
None, since you do a pred with no_grad and then immediately detach it

**Comment by sayakpaul** [comment]
Could we resolve the conflicts before merging? 

Apologies for the delay.

**Comment by AmericanPresidentJimmyCarter** [comment]
I will look into removing v-pred, cleaning up, and rebase this weekend.

**Comment by AmericanPresidentJimmyCarter** [comment]
Pushed.

**Review by sayakpaul — APPROVED** [review]
Thanks much! One final request is to add a small section of "DREAM" to the README so that the users are aware of it. Then we can merge.

**Review by yiyixuxu — APPROVED** [review]
looks good to me!
shall we merge? @sayakpaul

**Comment by yiyixuxu** [comment]
ohh 
@AmericanPresidentJimmyCarter do you want to add a section to README? otherwise, people won't notice this new technique we can use

**Comment by AmericanPresidentJimmyCarter** [comment]
Yeah, sorry, I forgot about this. I'll add the README later night.

**Comment by AmericanPresidentJimmyCarter** [comment]
@yiyixuxu OK it has README notes now.

**Comment by AmericanPresidentJimmyCarter** [comment]
`make quality` is doing weird things on main.

```
make quality
ruff check examples scripts src tests utils benchmarks setup.py
All checks passed!
ruff format --check examples scripts src tests utils benchmarks setup.py
Would reformat: examples/advanced_diffusion_training/train_dreambooth_lora_sd15_advanced.py
Would reformat: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py
Would reformat: examples/research_projects/multi_token_textual_inversion/textual_inversion.py
Would reformat: examples/research_projects/onnxruntime/textual_inversion/textual_inversion.py
Would reformat: examples/textual_inversion/textual_inversion.py
Would reformat: examples/textual_inversion/textual_inversion_sdxl.py
Would reformat: scripts/convert_svd_to_diffusers.py
Would reformat: src/diffusers/loaders/lora.py
Would reformat: src/diffusers/loaders/lora_conversion_utils.py
Would reformat: tests/models/autoencoders/test_models_vae.py
Would reformat: tests/pipelines/amused/test_amused.py
Would reformat: tests/pipelines/amused/test_amused_img2img.py
Would reformat: tests/pipelines/amused/test_amused_inpaint.py
Would reformat: utils/update_metadata.py
14 files would be reformatted, 876 files already formatted
make: *** [Makefile:44: quality] Error 1
```

**Comment by yiyixuxu** [comment]
can you install the most update version?
`pip install -e ".[quality]"`?

**Comment by AmericanPresidentJimmyCarter** [comment]
> can you install the most update version? `pip install -e ".[quality]"`?

That fixed it. Hopefully good to go.

**Comment by sayakpaul** [comment]
Thanks a ton @AmericanPresidentJimmyCarter!

**Comment by LiangbinXie** [comment]
@AmericanPresidentJimmyCarter Thanks for the implementation of DREAM. When I ran this method, I encountered a problem. Can you help me with that? Here is the reported issue.

image

It seems that the problem arose when the values of noisy_latent and target are set to (None, None). Is it a bug?

**Comment by AmericanPresidentJimmyCarter** [comment]
There is a bug in your code if you are using this function with two `None` tensors. As in the kwargs

```
    noisy_latents: torch.Tensor,
    target: torch.Tensor,
```

These should be not None.

**Comment by LiangbinXie** [comment]
@AmericanPresidentJimmyCarter From this code, you can find that the noisy_latents and target are set to None.
https://github.com/huggingface/diffusers/blob/main/src/diffusers/training_utils.py#L160

**Comment by AmericanPresidentJimmyCarter** [comment]
Ah, that's a bug I must have introduced during the refactor that no one caught. I'll PR a fix later.

**Comment by AmericanPresidentJimmyCarter** [comment]
https://github.com/huggingface/diffusers/pull/8302 should resolve it.

**Pull request merged**
